python_modules/dagster/dagster/core/definitions/repository.py

```diff
@@ -594,68 +621,129 @@ def from_list(cls, repository_definitions):
                     partition_sets[partition_set_def.name] = partition_set_def
             elif isinstance(definition, GraphDefinition):
                 coerced = definition.coerce_to_job()
-                pipelines[coerced.name] = coerced
+                if coerced.name in pipelines_and_jobs:
+                    raise DagsterInvalidDefinitionError(
+                        "Duplicate {target_type} definition found for {target}".format(
+                            target_type=definition.target_type, target=definition.describe_target()
+                        )
+                    )
+                pipelines_and_jobs[coerced.name] = coerced
             else:
                 check.failed(f"Unexpected repository entry {definition}")
 
         return CachingRepositoryData(
-            pipelines=pipelines,
+            pipelines_and_jobs=pipelines_and_jobs,
```

Comment:
```
pipelines = # all of the PipelineDefinitions in this list
jobs = # all of the JobDefinitions in this list

return CachingRepositoryData(
            pipelines=pipelines,
            jobs=jobs,
            partition_sets=partition_sets,
            schedules=schedules,
            sensors=sensors,
        )
```

Then, get_all_pipelines() (and related functions) should return everything that's in either the pipelines cache or the jobs cache.